experimental/tools/handshake-simulator/Simulation.cpp

```diff
@@ -730,120 +284,54 @@ HandshakeExecuter::HandshakeExecuter(
-        readyList.push_back(&op);
-      } else {
-        LLVM_DEBUG({
-          dbgs() << "EXECUTED: " << op << "\n";
-          for (auto out : op.getResults()) {
-            auto valueIt = valueMap.find(out);
-            if (valueIt != valueMap.end())
-              debugArg("OUT", out, valueMap[out], time);
-          }
-        });
-      }
-
-      for (mlir::Value out : scheduleList)
-        scheduleUses(readyList, valueMap, out);
-      continue;
-    }
+  ExecutableData execData {valueMap, memoryMap,       store,
+                          models,   internalDataMap, manager.currentCycle};
 
-    int64_t i = 0;
-    std::vector<Any> inValues(op.getNumOperands());
-    std::vector<Any> outValues(op.getNumResults());
-    bool reschedule = false;
-    LLVM_DEBUG(dbgs() << "OP: (" << op.getNumOperands() << "->"
-                      << op.getNumResults() << ")" << op << "\n");
-    time = 0;
-    for (mlir::Value in : op.getOperands()) {
-      if (valueMap.count(in) == 0) {
-        reschedule = true;
-        continue;
-      }
-      inValues[i] = valueMap[in];
-      time = std::max(time, timeMap[in]);
-      LLVM_DEBUG(debugArg("IN", in, inValues[i], timeMap[in]));
-      ++i;
-    }
-    if (reschedule) {
-      LLVM_DEBUG(dbgs() << "Rescheduling data...\n");
-      readyList.push_back(&op);
-      continue;
-    }
+  // Keeps track of operations that takes multiples cycles to not execute them
+  llvm::DenseMap<circt::Operation*, bool> cycleInternalChanged;
 
-    // Consume the inputs.
-    for (mlir::Value in : op.getOperands())
-      valueMap.erase(in);
-
-    ExecuteStrategy strat = ExecuteStrategy::Default;
-    LogicalResult res =
-        llvm::TypeSwitch<Operation *, LogicalResult>(&op)
-            .Case<mlir::arith::ConstantIndexOp, mlir::arith::ConstantIntOp,
-                  mlir::arith::AddIOp, mlir::arith::AddFOp, mlir::arith::CmpIOp,
-                  mlir::arith::CmpFOp, mlir::arith::SubIOp, mlir::arith::SubFOp,
-                  mlir::arith::MulIOp, mlir::arith::MulFOp,
-                  mlir::arith::DivSIOp, mlir::arith::DivUIOp,
-                  mlir::arith::DivFOp, mlir::arith::IndexCastOp,
-                  mlir::arith::ExtSIOp, mlir::arith::ExtUIOp,
-                  mlir::arith::XOrIOp>(
-                [&](auto op) {
-                  strat = ExecuteStrategy::Default;
-                  return execute(op, inValues, outValues);
-                })
-            .Case<circt::handshake::ReturnOp>([&](auto op) {
-              strat = ExecuteStrategy::Return;
-              return execute(op, inValues, outValues);
-            })
-            .Default([&](auto op) {
-              return op->emitOpError() << "Unknown operation";
-            });
-    LLVM_DEBUG(dbgs() << "EXECUTED: " << op << "\n");
-
-    if (res.failed()) {
-      successFlag = false;
-      return;
-    }
+  // Main simulation loop
+  while (true) {
+    // 1 cycle
+    while (manager.valueChangedThisCycle) {
```

Comment:
nit: consider making this a do-while for clarity, since we know we at least want to execute it once per execution of the outer loop

```cpp
do {
  manager.valueChangedThisCycle = false;
  // ...
} while (manager.valueChangedThisCycle)
```

This also allows you to remove the `manager.valueChangedThisCycle = true;` at the end of the outer loop.